### ISSUE TITLE
combined losses for segmentation task break distributed training

### ISSUE THREAD

**Issue by matteot11** [torchgeo/terratorch#982]
**Describe the issue**
I installed the most recent terratorch version from the main branch (commit `7cbf469f1c179860b81117512f873c6ff2b1ff5a`), to use the latest features (specifically, I was interested in combining multiple losses for semantic segmentation, like "ce"+"dice"). I used:
pip install git+https://github.com/IBM/terratorch.git

However, with this version installed, trainer.fit() hangs forever after the first train+val epoch. This only happens when **distributing training** over more than 1 GPUs and when using a **CombinedLoss**. It works when using a single loss, like "dice", instead.

**To Reproduce**
I attach a simple script for semantic segmentation trained with random data, to reproduce the issue:
```
import lightning as L
from terratorch.tasks import SemanticSegmentationTask
from torch.utils.data import Dataset, DataLoader
import torch

LOSS = "dice"  # works
# LOSS = {"ce": 1.0, "dice": 5.0}  # does not work


class DummyDataModule(L.LightningDataModule):
    def setup(self, stage) -> None:
        self.train_dataset = RandomDataset()
        self.val_dataset = RandomDataset()

    def train_dataloader(self) -> DataLoader:
        return DataLoader(
            dataset=self.train_dataset,
            batch_size=1,
            shuffle=True,
            num_workers=2,
            pin_memory=True,
            drop_last=True,
        )

    def val_dataloader(self) -> DataLoader:
        return DataLoader(
            dataset=self.val_dataset,
            batch_size=1,
            shuffle=False,
            num_workers=2,
            pin_memory=False,
            drop_last=False,
        )


class RandomDataset(Dataset):
    def __init__(self):
        self.imgs = [torch.rand((4, 4, 64, 64)) for _ in range(100)]
        self.masks = [torch.ones((64, 64)).long() for _ in range(100)]
        super().__init__()

    def __getitem__(self, idx):

        return {
            "image": self.imgs[idx],
            "mask": self.masks[idx],
        }

    def __len__(self):
        return len(self.imgs)


module = SemanticSegmentationTask(
    model_factory="EncoderDecoderFactory",
    model_args={
        "backbone": "prithvi_eo_v2_300",
        "backbone_pretrained": True,
        "backbone_img_size": 64,
        "backbone_num_frames": 4,
        "backbone_bands": ["BLUE", "GREEN", "RED", "NIR_NARROW"],
        "backbone_coords_encoding": [],
        "necks": [
            {"name": "SelectIndices", "indices": [5, 11, 17, 23]},
            {"name": "ReshapeTokensToImage", "effective_time_dim": 4},
            {"name": "LearnedInterpolateToPyramidal"},
        ],
        "decoder": "UNetDecoder",
        "decoder_channels": [512, 256, 128, 64],
        "head_dropout": 0.1,
        "num_classes": 2,
    },
    loss=LOSS,
    lr=1e-4,
    optimizer="AdamW",
    ignore_index=-1,
    freeze_backbone=False,
    freeze_decoder=False,
    plot_on_val=False,
)
datamodule = DummyDataModule()
trainer = L.Trainer(
    max_epochs=5,
    default_root_dir="./logs",
    num_sanity_val_steps=2,
    precision="bf16-mixed",
    log_every_n_steps=10,
    limit_val_batches=5,
    limit_train_batches=5,
)
trainer.fit(module, datamodule=datamodule)
```

**Screenshots or log output**
Lightning fit progress bar simply gets stuck at the end of the first epoch:
<img width="1446" height="21" alt="Image" src="https://github.com/user-attachments/assets/795e11aa-9539-467d-a0d4-858dfaba6366" />

**Expected behavior**
Training should work with CombinedLoss, too. 

**Deployment information**
Describe what you've deployed and how:
 - TerraTorch version: latest version from main branch (commit 7cbf469f1c179860b81117512f873c6ff2b1ff5a)
 - Installation source: pip
 - torch version: 2.9.0 (cu12)
 - devices: 2x NVIDIA TITAN RTX GPUs

I add @blumenstiel, since I noticed that the CombinedLoss feature has been recently added by a PR from you. Sorry if you are not the right person to ask.

**Comment by Joao-L-S-Almeida** [comment]
I could replicated it, with the only difference  that I used two A100. The training holds after the first epoch and this is the log I got:

```
INFO:albumentations.check_version:A new version of Albumentations is available: 2.0.8 (you have 1.4.10). Upgrade using: pip install --upgrade albumentations
INFO:root:Loaded weights for HLSBands.BLUE in position 0 of patch embed
INFO:root:Loaded weights for HLSBands.GREEN in position 1 of patch embed
INFO:root:Loaded weights for HLSBands.RED in position 2 of patch embed
INFO:root:Loaded weights for HLSBands.NIR_NARROW in position 3 of patch embed
WARNING:root:Decoder UNetDecoder does not have an `includes_head` attribute. Falling back to the value of the registry.
Using bfloat16 Automatic Mixed Precision (AMP)
Using default `ModelCheckpoint`. Consider installing `litmodels` package to enable `LitModelCheckpoint` for automatic upload to the Lightning model registry.
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
HPU available: False, using: 0 HPUs
Initializing distributed: GLOBAL_RANK: 0, MEMBER: 1/2
INFO:albumentations.check_version:A new version of Albumentations is available: 2.0.8 (you have 1.4.10). Upgrade using: pip install --upgrade albumentations
INFO:root:Loaded weights for HLSBands.BLUE in position 0 of patch embed
INFO:root:Loaded weights for HLSBands.GREEN in position 1 of patch embed
INFO:root:Loaded weights for HLSBands.RED in position 2 of patch embed
INFO:root:Loaded weights for HLSBands.NIR_NARROW in position 3 of patch embed
WARNING:root:Decoder UNetDecoder does not have an `includes_head` attribute. Falling back to the value of the registry.
Initializing distributed: GLOBAL_RANK: 1, MEMBER: 2/2
----------------------------------------------------------------------------------------------------
distributed_backend=nccl
All distributed processes registered. Starting with 2 processes
----------------------------------------------------------------------------------------------------

LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0,1]
LOCAL_RANK: 1 - CUDA_VISIBLE_DEVICES: [0,1]

  | Name          | Type             | Params | Mode 
-----------------------------------------------------------
0 | model         | PixelWiseModel   | 427 M  | train
1 | criterion     | CombinedLoss     | 0      | train
2 | train_metrics | MetricCollection | 0      | train
3 | val_metrics   | MetricCollection | 0      | train
4 | test_metrics  | ModuleList       | 0      | train
-----------------------------------------------------------
427 M     Trainable params
0         Non-trainable params
427 M     Total params
1,708.806 Total estimated model params size (MB)
650       Modules in train mode
0         Modules in eval mode
/dccstor/jlsa931/envs/dev/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py:4807: UserWarning: No device id is provided via `init_process_group` or `barrier `. Using the current device set by the user. 
  warnings.warn(  # warn only once
/dccstor/jlsa931/envs/dev/lib/python3.11/site-packages/torch/distributed/distributed_c10d.py:4807: UserWarning: No device id is provided via `init_process_group` or `barrier `. Using the current device set by the user. 
  warnings.warn(  # warn only once
/dccstor/jlsa931/envs/dev/lib/python3.11/site-packages/lightning/pytorch/loops/fit_loop.py:310: The number of training batches (5) is smaller than the logging interval Trainer(log_every_n_steps=10). Set a lower value for log_every_n_steps if you want to see logs for the training epoch.
/dccstor/jlsa931/envs/dev/lib/python3.11/site-packages/torch/autograd/graph.py:829: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
grad.sizes() = [1, 1, 1024], strides() = [66560, 1024, 1]
bucket_view.sizes() = [1, 1, 1024], strides() = [1024, 1024, 1] (Triggered internally at /pytorch/torch/csrc/distributed/c10d/reducer.cpp:334.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/dccstor/jlsa931/envs/dev/lib/python3.11/site-packages/torch/autograd/graph.py:829: UserWarning: Grad strides do not match bucket view strides. This may indicate grad was not created according to the gradient layout contract, or that the param's strides changed since DDP was constructed.  This is not an error, but may impair performance.
grad.sizes() = [1, 1, 1024], strides() = [66560, 1024, 1]
bucket_view.sizes() = [1, 1, 1024], strides() = [1024, 1024, 1] (Triggered internally at /pytorch/torch/csrc/distributed/c10d/reducer.cpp:334.)
  return Variable._execution_engine.run_backward(  # Calls into the C++ engine to run the backward pass
/dccstor/jlsa931/envs/dev/lib/python3.11/site-packages/torchmetrics/utilities/prints.py:43: UserWarning: The ``compute`` method of metric MulticlassAccuracy was called before the ``update`` method which may lead to errors, as metric states have not yet been updated.
  warnings.warn(*args, **kwargs)
```

**Comment by matteot11** [comment]
After some debugging, it looks like the problem originates from the logging logic here:
https://github.com/terrastackai/terratorch/blob/1306bff593ac2ca1c9837db0531b42cb957ec497/terratorch/tasks/loss_handler.py#L96-L104
Removing the explicit `on_epoch=True` and `on_step=True` arguments when logging individual losses prevents distributed training from freezing. Without these flags, pytorch-lightning falls back to the default logging behavior based on the hook context (i.e., `on_step` in `training_step`, `on_epoch` in `validation_step`, as described in the PL [docs](https://lightning.ai/docs/pytorch/stable/extensions/logging.html#automatic-logging)). However, I’m still unsure why logging the individual losses at both step and epoch level leads to sync issues in distributed training.

Moreover, I would ask why both flags were originally set for the individual loss logs, since the total loss already follows the default behavior here:
https://github.com/terrastackai/terratorch/blob/1306bff593ac2ca1c9837db0531b42cb957ec497/terratorch/tasks/loss_handler.py#L94

Is there a specific use case or task where logging individual losses at both step and epoch level is required? If not, I can open a PR to remove `on_epoch=True` and `on_step=True` from the individual loss logging calls.

**Comment by blumenstiel** [comment]
@matteot11 Thanks for debugging! I currently don't have the time to look into this in detail, sorry!
The log function is called with a single loss as well right? So I wonder why it only fails with the combined loss. In general, would be good to keep `on_epoch` and `on_step` as both are quite valuable to monitor the training. 

I wonder if Lightning wants to log it on every rank or if it correctly only logs on rank 0. But that should be the same for single losses as well...

**Comment by matteot11** [comment]
I apologize for not being very detailed. For context, here is the full `log_loss` method:
https://github.com/terrastackai/terratorch/blob/1306bff593ac2ca1c9837db0531b42cb957ec497/terratorch/tasks/loss_handler.py#L92-L104

**Case 1 — Single loss (e.g., "dice"):**
In this scenario, `loss_dict` only contains the "loss" key. This key is logged in the first `log_function` call, and the for loop is skipped (0 iterations).
In this first `log_function`, both `on_epoch` and `on_step` use Lightning's defaults, which results in the standard logging behavior and does not break distributed training.

**Case 2 — CombinedLoss (e.g., ["ce", "dice"]):**
Here, `loss_dict` contains "loss" plus individual components such as "ce" and "dice". The "loss" key is logged by the first `log_function` as expected and without any issue. The "ce" and "dice" keys are logged in the subsequent `log_function` calls inside the for loop. In these calls, both `on_epoch=True` and `on_step=True` are specified, which seems to trigger a synchronization issue in distributed training.

Additional Notes
Both `log_function` calls have `sync_dist=True`, so losses are reduced across processes prior to logging. But the issue here is caused by `on_epoch`/`on_step`, not by `sync_dist`.

In general, I think that the default Lightning logging behavior is preferred.
* During training, observing loss trends per step is useful and avoids waiting for epoch completion.
* During validation, we typically want the averaged loss over the full validation dataset to evaluate model improvement.

Thus, the default Lightning logging behavior (i.e., `on_step` for training loss, `on_epoch` for validation loss) seems more appropriate and I was wondering if there were any reasons not to follow it.

**Comment by bhansaliraunak** [comment]
@Joao-L-S-Almeida This looks like a distributed synchronisation deadlock during epoch-end metric reduction rather than a problem with the actual loss computation.

With a standard loss, only train_loss or val_loss are logged. 
With CombinedLoss, extra metrics such as 'ce' and 'dice' are also logged via LossHandler.log_loss() with sync_dist=True:
```
for loss_name, loss_value in all_losses.items():
    log_function(
        f"{self.loss_prefix}{loss_name}",
        loss_value,
        on_epoch=True,
        on_step=True,
        sync_dist=True,
    )
```

And so likely related to how CombinedLoss changes the loss object from a scalar Tensor to a dict[str, Tensor] upon additional loss components are logged under DDP.

One thing could be tested immediately:
if log_loss() modifies to temp disable sync_dist for component losses, and if the hang disappears, it could essentially prove that one of the component loss metrics is participating incorrectly in distributed reduction pipeline

**Comment by romeokienzler** [comment]
A fix is up in **#1219** (`fix-982-combined-loss-ddp-hang`): https://github.com/torchgeo/terratorch/pull/1219

Root cause matches @matteot11's and @bhansaliraunak's debugging — `LossHandler.log_loss` forced `on_step=True, on_epoch=True` (with `sync_dist=True`) on the individual component losses of a `CombinedLoss`, while the total loss used Lightning's defaults. The extra epoch-level synchronized reduction fired only for the components, desynchronising the DDP collectives across ranks and deadlocking at the first epoch boundary. Single losses skip that loop, which is why only `CombinedLoss` hung.

The PR logs component losses with the same settings as the total loss (Lightning picks `on_step`/`on_epoch` from the hook context, `sync_dist` kept) and adds `tests/test_loss_handler.py` guarding the logging contract. Reviews requested. Thanks all for the detailed diagnosis! 🙏

**Comment by romeokienzler** [comment]
Fixed by #1219, merged to `main` in dec7badd1.

**Root cause:** the loss handler logged the individual component losses of a `CombinedLoss` with `on_step=True`/`on_epoch=True` forced, while the total loss used Lightning's context defaults. At the first epoch boundary this fired an extra epoch-level *synced* reduction for the components but not the total, desynchronising the DDP collective calls across ranks and deadlocking training. A single (non-combined) loss skips the component loop, which is why only `CombinedLoss` runs hung.

**Fix:** component losses are now logged with the same settings as the total (`sync_dist=True`, `on_step`/`on_epoch` left to Lightning's hook context), plus regression tests in `tests/test_loss_handler.py`. Validated on the BlueVela cluster: unit 1603 passed, integration base set 16 passed.
